### PR TITLE
Add feature parity for s3 compliant providers

### PR DESCRIPTION
**Pull request by timuthy**
**What this PR does / why we need it**:
This PR adds support for S3 compatible providers (see #431) when configuration is passed via secrets.

**Release note**:

```improvement operator
None
```

### CODE REVIEW

**Comment by ishan16696** [comment]
@timuthy Can you resolve the merge conflicts ?

File: pkg/snapstore/s3_snapstore.go [hunk 1]
```diff
@@ -184,24 +188,41 @@ func readAWSCredentialFromDir(dirname string) (*awsCredentials, error) {
...
 			}
 			awsConfig.SecretAccessKey = string(data)
+		case "endpoint":
+			data, err := ioutil.ReadFile(dirname + "/endpoint")
+			if err != nil {
+				return nil, err
+			}
+			awsConfig.Endpoint = pointer.StringPtr(string(data))
+		case "s3ForcePathStyle":
+			data, err := ioutil.ReadFile(dirname + "/s3ForcePathStyle")
```
Comment (ishan16696) [inline pkg/snapstore/s3_snapstore.go:217]:
```suggestion
			data, err := os.ReadFile(dirname + "/s3ForcePathStyle")
```

Comment (ishan16696) [reply pkg/snapstore/s3_snapstore.go:217]: as now we have move to Go `v1.17`

**Review by ishan16696 — CHANGES_REQUESTED** [review]

**Comment by timuthy** [comment]
Thanks for the review @ishan16696. PTAL.

File: pkg/snapstore/s3_snapstore.go [hunk 2]
```diff
@@ -141,8 +143,10 @@ func readAWSCredentialsJSONFile(filename string) (session.Options, error) {
 
 	return session.Options{
 		Config: aws.Config{
-			Credentials: credentials.NewStaticCredentials(awsConfig.AccessKeyID, awsConfig.SecretAccessKey, ""),
-			Region:      pointer.StringPtr(awsConfig.Region),
+			Credentials:      credentials.NewStaticCredentials(awsConfig.AccessKeyID, awsConfig.SecretAccessKey, ""),
+			Region:           pointer.StringPtr(awsConfig.Region),
+			Endpoint:         awsConfig.Endpoint,
```
Comment (ishan16696) [inline pkg/snapstore/s3_snapstore.go:148]: What happen when there is no `Endpoint` configure as `Endpoint` used only for S3 compatible providers? Have you tested that scenario?

Comment (timuthy) [reply pkg/snapstore/s3_snapstore.go:148]: Nothing happens, the Endpoint just remains `nil`. This is the same as not setting it at all, like we do today https://github.com/gardener/etcd-backup-restore/blob/c48c9973106837a87a652da131cbd6d5bf19b32a/pkg/snapstore/s3_snapstore.go#L78.

**Pull request merged**